### PR TITLE
[metrics] Redesign metrics to 3 layers

### PR DESCRIPTION
**Pull request by acelyc111**
## Proposed changes
In order to support more fine-grained metrics, like table or tablet level metrics, I redesign metrics to 3 layers:
**MetricRegistry - MetricEntity - Metrics**
The roles in the new design:
- MetricRegistry
   the register center
- MetricEntity
   the enity registered on MetricRegistry.
   Generally a MetricRegistry can be registered on serveral MetricEntities, each of MetricEntity is an independent entity, such as server, disk_devices, data_directories, thrift clients and servers, and so on.
- Metric
   metrics of an entity.
   Such as fragment_requests_total on server entity, disk_bytes_read on a disk_device entity, thrift_opened_clients on a thrift_client entity.
- MetricPrototype
   The type of a Metric.
   MetricPrototype is a global variable, can be shared by the same metrics across different MetricEntities.
   For example, different tablets can share the same MetricPrototype, but own it's own Metrics which have different values.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

All metrics HTTP interfaces have the same format of output as before.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have create an issue on [Doris's issues](https://github.com/apache/incubator-doris/issues), and have described the bug/feature there in detail
- [ ] Commit messages in my PR start with the related issues ID, like "#4071 Add pull request template to doris project"
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/util/doris_metrics.h [hunk 1]
```diff
@@ -28,163 +28,152 @@
...
+    //METRIC_DEFINE_INT_COUNTER(load_rows_total, MetricUnit::ROWS);
+    //METRIC_DEFINE_INT_COUNTER(load_bytes_total, MetricUnit::BYTES);
+
+    //IntGaugeMetricsMap disks_total_capacity;
+    //IntGaugeMetricsMap disks_avail_capacity;
+    //IntGaugeMetricsMap disks_data_used_capacity;
+    //IntGaugeMetricsMap disks_state;
+
+    IntCounter fragment_requests_total;
+	IntCounter fragment_request_duration_us;
```
Comment (morningman) [inline be/src/util/doris_metrics.h:52]: Use 4 space instead of tab

File: be/src/util/doris_metrics.h [hunk 2]
```diff
@@ -28,163 +28,152 @@
...
-    METRIC_DEFINE_INT_COUNTER(publish_task_failed_total, MetricUnit::REQUESTS);
-
-    METRIC_DEFINE_INT_COUNTER(meta_write_request_total, MetricUnit::REQUESTS);
-    METRIC_DEFINE_INT_COUNTER(meta_write_request_duration_us, MetricUnit::MICROSECONDS);
-    METRIC_DEFINE_INT_COUNTER(meta_read_request_total, MetricUnit::REQUESTS);
-    METRIC_DEFINE_INT_COUNTER(meta_read_request_duration_us, MetricUnit::MICROSECONDS);
+    //METRIC_DEFINE_INT_COUNTER(load_rows_total, MetricUnit::ROWS);
+    //METRIC_DEFINE_INT_COUNTER(load_bytes_total, MetricUnit::BYTES);
+
+    //IntGaugeMetricsMap disks_total_capacity;
```
Comment (morningman) [inline be/src/util/doris_metrics.h:46]: Remove the unused code

**Comment by morningman** [comment]
I took a cursory look at the code and found no problems.
This CL is too big and involves too many modules...
Have you run it in your product env yet？

**Comment by acelyc111** [comment]
> I took a cursory look at the code and found no problems.
> This CL is too big and involves too many modules...
> Have you run it in your product env yet？

Yes, I've run it in our product env, it works well.

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by morningman** [comment]
I've test it in our env and it looks good to me.

File: be/src/util/metrics.h [hunk 3]
```diff
@@ -193,230 +174,163 @@ class CoreLocalCounter : public Metric {
...
-    std::string value;
+    std::string description;
+    std::string group_name;
+    Labels labels;
+};
 
-    MetricLabel() { }
-    MetricLabel(const std::string& name_, const std::string& value_) :name(name_), value(value_) {
-    }
+#define DEFINE_METRIC(name, type, unit, desc, group, labels, core)      \
```
Comment (chaoyli) [inline be/src/util/metrics.h:245]: may be used DEFINE_METRIC_PROTOTYPE better

File: be/src/util/metrics.h [hunk 4]
```diff
@@ -193,230 +174,163 @@ class CoreLocalCounter : public Metric {
...
-      : LockSimpleMetric<T>(MetricType::GAUGE, unit) {}
-    virtual ~LockGauge() { }
+    LockGauge() : LockSimpleMetric<T>() {}
+    virtual ~LockGauge() {}
 };
 
-// one key-value pair used to
-struct MetricLabel {
+using Labels = std::unordered_map<std::string, std::string>;
+struct MetricPrototype {
```
Comment (chaoyli) [inline be/src/util/metrics.h:216]: What's the purpose of MetricPrototype?
Under which circumstance, MetricPrototype can be used to split the Metric.

File: be/src/util/metrics.h [hunk 5]
```diff
@@ -193,230 +174,163 @@ class CoreLocalCounter : public Metric {
...
-        }
-        return ss.str();
+struct MetricPrototypeEqualTo {
+    bool operator()(const MetricPrototype* first, const MetricPrototype* second) const {
+        return first->group_name == second->group_name && first->name == second->name;
     }
 };
 
-class MetricCollector;
+using MetricByType = std::unordered_map<const MetricPrototype*, Metric*, MetricPrototypeHash, MetricPrototypeEqualTo>;
```
Comment (chaoyli) [inline be/src/util/metrics.h:285]: MetricMap is better name or not?

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/util/metrics.h (hunk 5 above)
Comment (acelyc111) [reply be/src/util/metrics.h]: MetricMap would be better

**Review by chaoyli — APPROVED** [review]

**Pull request merged**